### DISCUSSION TITLE
Elsa with Angular Frontend

### DISCUSSION THREAD

**Discussion by CansuGoren** [elsa-workflows/elsa-core#1195] [Q&A]
We are decided to use Elsa  with my current company.
But I have some question and couldn’t find anywhere the answer.

- First of all, how can I tigger my workflow from the frontend developed with Angular, is it possible?
- Is there any pros or cons between trigger workflow from frontend and http endpoint request  and received a signal, which method should I prefer?
- Another question is ıs there any restriction about contract first project to use elsa workflow?

Thank you in advance.

**Comment by sfmskywalker** [comment]
Hi Cansu,

1. That’s certainly possible. One way is to invoke Elsa’s REST API at `/v1/workflows/{workflowDefinitionId}/execute`. Another way is to start your workflow with the **HTTP Endpoint** activity and make requests to that one instead. Yet another approach is implementing a custom API controller from which you execute workflows.
2. It depends on your specific use cases, but one approach isn’t necessarily better than the other. I would probably start out with invoking the REST API if my front-end already knows about which workflow to execute. It’s practically the same as adding an HTTP Endpoint to the workflow and invoking that one directly. It doesn’t really matter which way you go.
3. No restrictions, use Elsa Workflows in any way you see fit.

**Reply by CansuGoren** [reply]
Hi Sipke,

Thank you both your rapidly answer my question and introduce to us Elsa. I have another issue now,
After register documents and information control, If everything is ok, I want to resume my workflow and I thought that I could do that using send signal.
How can I send a signal from another application( like running with angular js frontend), is it possible?
Thank you in advance.

**Reply by prven-affirma** [reply]
> Hi Cansu,
> 
> 1. That’s certainly possible. One way is to invoke Elsa’s REST API at `/v1/workflows/{workflowDefinitionId}/execute`. Another way is to start your workflow with the **HTTP Endpoint** activity and make requests to that one instead. Yet another approach is implementing a custom API controller from which you execute workflows.
> 2. It depends on your specific use cases, but one approach isn’t necessarily better than the other. I would probably start out with invoking the REST API if my front-end already knows about which workflow to execute. It’s practically the same as adding an HTTP Endpoint to the workflow and invoking that one directly. It doesn’t really matter which way you go.
> 3. No restrictions, use Elsa Workflows in any way you see fit.

We are in the same boat as well. When I tried calling a sample workflow using REST API (POSTMAN), I keep getting 400 Bad Request. Can you please suggest what might be wrong here? But when I make any GET calls, those API calls works as expected.

![image](https://user-images.githubusercontent.com/38470789/124207774-1b6f5800-da9b-11eb-8237-8fd07427ae92.png)

**Reply by sfmskywalker** [reply]
Make sure to send along the content-type header with application/json. When using Postman, the easiest way is to set Body to “raw” and then select the JSON content type.

**Reply by CansuGoren** [reply]
In my case I added services.AddHttpActivities() twice in Startup.cs
Make sure you add this service or don't add it twice :)

**Reply by prven-affirma** [reply]
> Make sure to send along the content-type header with application/json. When using Postman, the easiest way is to set Body to “raw” and then select the JSON content type.

I added both (the Header & Body raw) and still getting the same 400.

![image](https://user-images.githubusercontent.com/38470789/124301516-ce2fcc80-db14-11eb-8392-97cd6d5ca98c.png)

**Reply by sfmskywalker** [reply]
What happens if you include an empty JSON object? `{}`

**Reply by prven-affirma** [reply]
No difference. Please let me know if there is a way to DM you with the URL.

**Reply by sfmskywalker** [reply]
Sure, I’m on Discord, DM me from there. Als send your project source code if possible so I can reproduce.

**Reply by prven-affirma** [reply]
Sent an email to sipkeschoorstra@outlook.com with the source code. Our team is blocked currently because of this. Appreciate all your help. Thank you!

**Reply by sfmskywalker** [reply]
I see now. A couple of notes:

1. You're using the 2.0 release that doesn't yet expose the updated API endpoints you're trying to use.
2. When you try to execute a workflow using the API that starts with a trigger (like `HTTP Endpoint`), you need to make sure to provide the expected input.
3. More importantly, you don't have to invoke workflows that start with a trigger yourself.

In the case of your Hello World workflow, which starts with an HTTP Endpoint, all you need to do to trigger that one is to make an HTTP GET request to https://localhost:5001/trigger.

If you **DO** want to execute a workflow using the API endpoint, that works too of course - simply get rid of the trigger from your workflow. In the case of your Hello World workflow, simply delete the HTTP Endpoint activity, then it works.

Or, taking one of your sample workflows, **testsamplecalled**, it works when you do this:

```bash
curl --location --request POST 'https://localhost:5001/v1/workflows/344bbf755db64d85ab085827ee5a6296/execute' \
--header 'Content-Type: application/json' \
--data-raw '{}'
```

Response:

```json
{
    "$id": "1",
    "executed": true,
    "activityId": "660f0223-25a5-4772-b304-38f5d37f0679",
    "workflowInstance": {
        "$id": "2",
        "definitionId": "344bbf755db64d85ab085827ee5a6296",
        "version": 2,
        "workflowStatus": "Finished",
        "correlationId": "c8df57fb8f354397b3a1347faab3c889",
        "createdAt": "2021-07-04T18:26:57.0851125Z",
        "lastExecutedAt": "2021-07-04T18:27:02.7835662Z",
        "finishedAt": "2021-07-04T18:27:02.7843466Z",
        "variables": {
            "$id": "3",
            "data": {}
        },
        "activityData": {
            "$id": "4",
            "660f0223-25a5-4772-b304-38f5d37f0679": {
                "$id": "5",
                "_Lifecycle": {
                    "$id": "6",
                    "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
                    "executing": true,
                    "executed": true
                }
            }
        },
        "blockingActivities": [],
        "scheduledActivities": [],
        "scopes": [],
        "lastExecutedActivityId": "660f0223-25a5-4772-b304-38f5d37f0679",
        "id": "9c56ca1581fb4684bdca500a4198eada"
    }
}
```

To summarize, make sure to:

1. Update to the latest PREVIEW packages of Elsa
2. Don't use the API endpoints to execute workflows that start with a trigger (you could make that work, but there's no point to do so).

**Important!**
While trying out your code I updated to the preview packages, which includes newer EF Core migrations that got executed on your database. Let me know if this is a problem (should you want to stay with the Elsa 2.0 release rather than the 2.1 preview packages), then I will roll back the migration.

Note however that 2.1 preview is close to being released (later this week), so I would recommend using the 2.1 preview packages to take advantage of all the fixes and enhancements it provides.

Let me know.

**Reply by prven-affirma** [reply]
Can you please let me know where I can find the Elsa 2.1 previous package? In Nuget package manager I do not see 2.1 Preview. 

![image](https://user-images.githubusercontent.com/38470789/124408875-0a715180-dcfc-11eb-985e-5b82f5f27f60.png)

**Reply by prven-affirma** [reply]
Never mind, found it. Thank you for all your support. We will give it a try and let you know.

**Comment by abdallahwishah** [comment]
i have aissue 
i used this lisk https://aspnetzero.com/blog/integrating-elsa-with-aspnet-zero-angular
to integrating ELSA with ASP.NET Zero (Angular)
every thing was working fine 
but when i use this way 
![image](https://github.com/elsa-workflows/elsa-core/assets/84712712/3b02a2ab-7bf3-45cf-8b36-8ce50c6cded3)
to send token 
all basic header removed 
befor:
![image](https://github.com/elsa-workflows/elsa-core/assets/84712712/e0269587-e80b-4b31-83f6-558e5bfd68f0)
after use :

![image](https://github.com/elsa-workflows/elsa-core/assets/84712712/9186ada8-715b-460e-a844-ff21a4e2d63b)

and i replaced
this:	request.headers = { Authorization: `Bearer ${token}` }; 
to 
this   request.headers = {
                                ...request.headers,
                                Authorization: 'Bearer secret-token',
                            };

**Comment by abdallahwishah** [comment]
@sfmskywalker
